### PR TITLE
feat(core): improved lock context propagation support

### PR DESCRIPTION
**Pull request by cmark**
Fixes collection versioning issue when the underlying child resource has synchronizable content to other toolings and requires a lock to write into the repository (but was unable to acquire it due to missing parent context).

### CODE REVIEW

**Review by apeteri — APPROVED** [review]
LGTM!

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/IOperationLockManager.java [hunk 1]
```diff
@@ -17,65 +17,80 @@
...
+	 *            the lock context (may not be {@code null})
+	 * @param firstTarget
+	 *            the first (or only) target to unlock (may not be {@code null})
+	 * @param restTargets
+	 *            subsequent targets to unlock (may not be {@code null}; individual elements may not be {@code null})
+	 * @throws IllegalArgumentException
+	 *             when one or more locks for the given targets could not be released for some reason
 	 */
-	void unlock(DatastoreLockContext context, DatastoreLockTarget firstTarget, DatastoreLockTarget... restTargets) throws IllegalArgumentException;
+	void unlock(DatastoreLockContext context, Lockable firstTarget, Lockable... restTargets) throws IllegalArgumentException;
```
Comment (nagyo) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/IOperationLockManager.java]: The lock() counterpart has a `lockable` and a `lockables` parameter. What do you think about using something similar here, just for the sake of consistency? Or `target` and `targets` or whatever suits the purpose best. The `restTargets` tripped me to think about REST for some reason :dumbo:

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/version/VersioningConfiguration.java [hunk 2]
```diff
@@ -35,20 +35,23 @@ public final class VersioningConfiguration {
...
 	
 	public VersioningConfiguration(
 			String user,
 			ResourceURI resource,
 			String version, 
 			String description,
 			LocalDate effectiveTime,
-			boolean force) {
+			boolean force,
+			boolean versioningAsChildResource) {
```
Comment (nagyo) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/version/VersioningConfiguration.java]: The verb `versioning` might be redundant here. This is the __Versioning__ Configuration, if the parameter would be e.g. `asChildResource` I think the operation itself is implied.

**Review by nagyo — APPROVED** [review]
LGTM, with some small remarks

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/IOperationLockManager.java (hunk 1 above)
Comment (cmark) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/IOperationLockManager.java]: :nod:

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/version/VersioningConfiguration.java (hunk 2 above)
Comment (cmark) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/version/VersioningConfiguration.java]: Changing it to `childResource` as isAsChildResource accessor is a bit weird 😄.

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/IOperationLockManager.java (hunk 1 above)
Comment (cmark) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/IOperationLockManager.java]: Done

**Pull request merged**